content/articles/get-started-with-a-lexical-analyzer/index.md

```diff
@@ -0,0 +1,534 @@
+
+**Subroutines**: C codes that state additional user-defined functions executed when the patterns are found or not found.
+They can also contain other functions that the program may execute on the run.
+These are optional.
+
+The three sections are divided by delimiters (`%%`) as shown below:
+```
+1. Definitions section:
+
+%{    contains C code, usually consisting of the libraries
+        #define external global declarations that one wants to be part of the code for y.tab.c
+%}
+
+The file can include optional macro definitions for regular expressions.
+
+%%
+
+2. Rules section:
+
+regular expression        { c code to be executed when this R.E. is matched against the source. }
+another regular expression { another c code based on this RegEX }
+%%
+
+3. User Sub-routines:
+
+Additional code here
+```
+### Create a Lex file and run it
+
+Let us create a Lex analyzer that counts the number of words in a file and their total size.
+
+- Create a root folder named 'flex-projects'.
+
+- In it, create a file named `counter.l`
+
+- Open it and do the following: Add lines of code that state the program declarations as follows:
+```c
+/* Definitions */
+%{
+int lines=1, words=0, lowercase_letters=0, uppercase_letters=0, numbers=0, special_characters=0, totals=0, size=0;
+%}
+
+/* Delimiters: These separate the definitions from the Rules and actions section */
+%%
+```
+
+- Add the rules and actions. Uppercase and lowercase letters, numbers, special characters, letters, and line instances will all be checked.Check it out in the code below:
+```c
+\n { lines++; words++;} /* Adds to the lines and words variables a value of 1 when one goes to a new line */
+ [\t ' '] words++; /* Adds a value to the words variable when one moves from one word to another (words are separated by a space) */
+ [A-Z] uppercase_letters++;
+ [a-z] lowercase_letters++;
+ [0-9] numbers++;
+ [$&+,:;=?@#|'<>.^*()%!-] special_characters++;
+ 
+ /* Separate the Rules and Action Section from the User sub-routines */
+ %%
+```
+
+- In the User sub-routines section, add the main program that prints the totals. Check it out in the line of code below:
+```c
+int yywrap()
+{
+
+}
+
+int main()
+{
+yyin= fopen("test.txt", "r"); /* Open the 'test.txt' file in a read only fomart */
+yylex();
+totals= special_characters + numbers + uppercase_letters + lowercase_letters;
+size = (totals * 1); /* The total size in bytes is equal to the total number of characters multiplied by one byte since each is one byte */
+
+/* Prints the output */
+printf("This file contains the following:");
+printf("\n\t%d lines", lines);
+printf("\n\t%d words", words);
+printf("\n\t%d Lowercase letters", lowercase_letters);
+printf("\n\t%d Uppercase letters", uppercase_letters);
+printf("\n\t%d digits", numbers);
+printf("\n\t%d special characters", special_characters);
+printf("\n\t The total size of the file characters in bytes is: %d bytes.\n", size);
+}
```

Comment:
Can you format these?

Reply:
when you add indentation on lexical files, the whole app breaks. They are very sensitive to structure and indentation

Reply:
Got it. Thanks for clarification.